src/coreclr/jit/valuenum.cpp

```diff
@@ -7768,82 +7777,80 @@ void Compiler::fgValueNumberAssignment(GenTreeOp* tree)
+                                newFirstFieldValueVN = vnStore->VNApplySelectorsAssignTypeCoerce(storeVal, indType);
+                            }
+                            else
+                            {
+                                // Construct the ValueNumber for fldMap[obj/offset]. This (struct)
+                                // map represents the specific field we're looking to store to.
+                                ValueNum firstFieldValueVN =
+                                    vnStore->VNForMapSelect(VNK_Liberal, firstFieldType, fldMapVN,
+                                                            firstFieldValueSelectorVN);
+
+                                // Construct the maps updating the rest of the fields in the sequence.
+                                newFirstFieldValueVN =
+                                    vnStore->VNApplySelectorsAssign(VNK_Liberal, firstFieldValueVN, fldSeq->m_next,
+                                                                    storeVal, indType);
                             }
 
-                            // From which we can construct the new ValueNumber for 'fldMap at normVal'
-                            newFldMapVN =
-                                vnStore->VNForMapStore(vnStore->TypeOfVN(fldMapVN), fldMapVN, normVal, storeVal);
+                            // Finally, construct the new field map...
+                            ValueNum newFldMapVN =
+                                vnStore->VNForMapStore(vnStore->TypeOfVN(fldMapVN), fldMapVN, firstFieldValueSelectorVN,
+                                                       newFirstFieldValueVN);
+
+                            // ...and a new value for the heap.
+                            newHeapVN = vnStore->VNForMapStore(TYP_REF, fgCurMemoryVN[GcHeap], firstFieldSelectorVN,
+                                                               newFldMapVN);
                         }
                         else
                         {
-                            // plain static field
-
-                            // Now get rid of any remaining struct field dereferences. (if they exist)
-                            if (fldSeq->m_next)
-                            {
-                                storeVal = vnStore->VNApplySelectorsAssign(VNK_Liberal, fldMapVN, fldSeq->m_next,
-                                                                           storeVal, indType);
-                            }
-
-                            newFldMapVN = vnStore->VNApplySelectorsAssign(VNK_Liberal, fgCurMemoryVN[GcHeap], fldSeq,
-                                                                          storeVal, indType);
+                            // Plain static field.
+                            newHeapVN = vnStore->VNApplySelectorsAssign(VNK_Liberal, fgCurMemoryVN[GcHeap], fldSeq,
+                                                                        storeVal, indType);
```

Comment:
Is this dead code?

Reply:
As far as I know, yes, because we use `CLS_VAR` for these cases.

(I would not be surprised if it is reachable under some exotic circumstances of course)

Reply:
I don't think so, `IsFieldAddr` can never return true without assigning one of the out args. Maybe at some point code like the following would hit this path:
```csharp
public class Program
{
    public static S s_s;
    public static int s_i;

    public static void Main()
    {
        s_s = s_s;
        Foo();
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static unsafe void Foo()
    {
        Bar(ref s_i);
    }

    public static void Bar(ref int i) => i = 10;
}
```
but currently we do not manage to VN this:
```scala
***** BB01, STMT00002(before)
N004 (  6, 14) [000006] -A--G-------              ▌  ASG       int   
N002 (  4, 12) [000005] *------N----              ├──▌  IND       int   
N001 (  2, 10) [000003] I-----------              │  └──▌  CNS_INT(h) byref  0x7ffa80d505b8 static Fseq[s_i]
N003 (  1,  1) [000004] ------------              └──▌  CNS_INT   int    10

N001 [000003]   CNS_INT(h) 0x7ffa80d505b8 static Fseq[s_i] => $c0 {Hnd const: 0x00007FFA80D505B8}
N003 [000004]   CNS_INT   10 => $41 {IntCns 10}
  fgCurMemoryVN[GcHeap] assigned for assign-of-IND at [000006] to VN: $81.
N004 [000006]   ASG       => $VN.Void

***** BB01, STMT00002(after)
N004 (  6, 14) [000006] -A--G-------              ▌  ASG       int    $VN.Void
N002 (  4, 12) [000005] *------N----              ├──▌  IND       int    $41
N001 (  2, 10) [000003] I-----------              │  └──▌  CNS_INT(h) byref  0x7ffa80d505b8 static Fseq[s_i] $c0
N003 (  1,  1) [000004] ------------              └──▌  CNS_INT   int    10 $41
```